### ISSUE TITLE
Safe mode with sparse linear solver other than MA57

### ISSUE THREAD

**Issue by nychiang** [LLNL/hiop#452]
The current workflow will replace the sparse linear solver to MA57, even though we set `linear_solver_sparse` to other solvers.

This is due to the fact that `linsol_safe_mode_on` is set to `true` by default in [hiopAlgFilterIPM](https://github.com/LLNL/hiop/blob/083914f0f619fe511abfa3c9031c0926e1a0df02/src/Optimization/hiopAlgFilterIPM.cpp#L1829-L1876) and hence we set `kkt->set_safe_mode(linsol_safe_mode_on);` 

Later `determineAndCreateLinsys` will use `MA57` since `safe_mode` is `true` and ignore the other solvers, e.g. pardiso or strumpack, see [hiopKKTLinSysSparse.cpp](https://github.com/LLNL/hiop/blob/083914f0f619fe511abfa3c9031c0926e1a0df02/src/Optimization/hiopKKTLinSysSparse.cpp#L567-L589)

@cnpetra

**Comment by cnpetra** [comment]
looks like you're running with `linsol_mode` set to `stable`. I could not reproduce with `speculative`

**Comment by nychiang** [comment]
yes, you are correct. I am using the default setting which sets `linsol_mode` to `stable`.
Probably we need a safeguard which set `linsol_mode` to `speculative` when CPU solver other than MA57 is given.

**Comment by nychiang** [comment]
@cnpetra  This problem still exists. In short, if we set `linsol_mode=stable` (default option), sparse solver only uses `MA57`. That is, hiop option `linear_solver_sparse` is ignored. For example, Pardiso won't work under settings `linsol_mode=stable` and `linear_solver_sparse=pardiso`.

**Comment by cnpetra** [comment]
as per our discussion

1. `linsol_mode=stable` should instantiate the what `linear_solver_sparse` suggests.

2. under speculative linsol mode, the backup should be MA57.

**Comment by huojiangming** [comment]
Cross-referencing three reports that land on this same code, all checked against `develop`
(fc0e881) in a build **without** COINHSL:

* #778 — the other half of what you describe: with `HIOP_USE_COINHSL` off, the safe-mode
  branch of `hiopKKTLinSysCompressedSparseXDYcYd::determineAndCreateLinsys` has no solver
  at all. It is an `assert(false && "HiOp was not built with the safe(r) sparse linear
  solver MA57 …")` followed by `return nullptr`, so a release build returns a null solver
  rather than falling back.
* #790 — the assert just above it requires `compute_mode` to be `hybrid`/`cpu`/`auto`, so
  entering safe mode on a GPU run is a failure rather than a slower solve.
* #798 — `switch_to_safer_KKT` sets `fact_acceptor` to `inertia_correction`, which the GPU
  sparse solvers do not support, so the two interact.

Posting here rather than opening anything new; if you would rather have them as one report
on this issue, say so and I will close them.
